From b90c05bc8ca5d456053262942c530dcbac4dcd9a Mon Sep 17 00:00:00 2001
From: Scott Talbert <swt@techie.net>
Date: Thu, 26 Feb 2015 23:21:59 -0500
Subject: [PATCH] Fix wxGtkPrintDialog when created standalone

In fixing #14731, the GtkPrintOperation creation was moved from
wxGtkNativePrintData to within wxGtkPrinter::Print().  However,
wxGtkPrintDialog can also be created directly (ie, not using wxGtkPrinter)
so it needs a GtkPrintOperation too.  Move the GtkPrintOperation creation
into wxGtkPrintDialog.

Fixes #16821.
---
 src/gtk/print.cpp | 24 +++++++++++++++++++-----
 1 file changed, 19 insertions(+), 5 deletions(-)

diff --git a/src/gtk/print.cpp b/src/gtk/print.cpp
index 2d70a1b..0082ac4 100644
--- a/src/gtk/print.cpp
+++ b/src/gtk/print.cpp
@@ -612,6 +612,11 @@ wxGtkPrintDialog::wxGtkPrintDialog( wxWindow *parent, wxPrintDialogData *data )
 
     m_parent = parent;
     SetShowDialog(true);
+
+    wxPrintData printData = m_printDialogData.GetPrintData();
+    wxGtkPrintNativeData *native =
+      (wxGtkPrintNativeData*) printData.GetNativeData();
+    native->SetPrintJob(gtk_print_operation_new());
 }
 
 wxGtkPrintDialog::wxGtkPrintDialog( wxWindow *parent, wxPrintData *data )
@@ -625,11 +630,22 @@ wxGtkPrintDialog::wxGtkPrintDialog( wxWindow *parent, wxPrintData *data )
 
     m_parent = parent;
     SetShowDialog(true);
+
+    wxPrintData printData = m_printDialogData.GetPrintData();
+    wxGtkPrintNativeData *native =
+      (wxGtkPrintNativeData*) printData.GetNativeData();
+    native->SetPrintJob(gtk_print_operation_new());
 }
 
 
 wxGtkPrintDialog::~wxGtkPrintDialog()
 {
+    wxPrintData data = m_printDialogData.GetPrintData();
+    wxGtkPrintNativeData *native =
+      (wxGtkPrintNativeData*) data.GetNativeData();
+    GtkPrintOperation * const printOp = native->GetPrintJob();
+    g_object_unref(printOp);
+    native->SetPrintJob(NULL);
 }
 
 // This is called even if we actually don't want the dialog to appear.
@@ -921,10 +937,9 @@ bool wxGtkPrinter::Print(wxWindow *parent, wxPrintout *printout, bool prompt )
     wxPrintData printdata = GetPrintDialogData().GetPrintData();
     wxGtkPrintNativeData *native = (wxGtkPrintNativeData*) printdata.GetNativeData();
 
-    wxGtkObject<GtkPrintOperation> printOp(gtk_print_operation_new());
-    native->SetPrintJob(printOp);
-    wxON_BLOCK_EXIT_OBJ1(*native, wxGtkPrintNativeData::SetPrintJob,
-                         static_cast<GtkPrintOperation*>(NULL));
+    // wxGtkPrintDialog needs to be created first as it creates the PrintOp
+    wxGtkPrintDialog dialog( parent, &m_printDialogData );
+    GtkPrintOperation * const printOp = native->GetPrintJob();
 
     wxPrinterToGtkData dataToSend;
     dataToSend.printer = this;
@@ -937,7 +952,6 @@ bool wxGtkPrinter::Print(wxWindow *parent, wxPrintout *printout, bool prompt )
 
     // This is used to setup the DC and
     // show the dialog if desired
-    wxGtkPrintDialog dialog( parent, &m_printDialogData );
     dialog.SetPrintDC(m_dc);
     dialog.SetShowDialog(prompt);
 
-- 
2.1.0

